File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *   val: number
 *   left: TreeNode | null
 *   right: TreeNode | null
 *   constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 *   }
 * }
 */

function maxPathSum(root: TreeNode | null): number {
  let ans: number = root.val
  
  function dfs(node: TreeNode | null): number {
    if (!node) return 0
    const leftMax: number = Math.max(dfs(node.left), 0)
    const rightMax: number = Math.max(dfs(node.right), 0)
    
    ans = Math.max(ans, node.val + leftMax + rightMax)
    return node.val + Math.max(leftMax, rightMax)
  }
  
  dfs(root)
  return ans
}

File: 0124-binary-tree-maximum-path-sum/0124-binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *   int val;
 *   TreeNode left;
 *   TreeNode right;
 *   TreeNode() {}
 *   TreeNode(int val) { this.val = val; }
 *   TreeNode(int val, TreeNode left, TreeNode right) {
 *     this.val = val;
 *     this.left = left;
 *     this.right = right;
 *   }
 * }
 */
class Solution {
  public int dfs(TreeNode node, int[] ans) {
    if (node == null) return 0;
    int leftMax = Math.max(dfs(node.left, ans), 0);
    int rightMax = Math.max(dfs(node.right, ans), 0);
    ans[0] = Math.max(ans[0], node.val + leftMax + rightMax);
    return node.val + Math.max(leftMax, rightMax);
  }
  
  public int maxPathSum(TreeNode root) {
    int[] ans = new int[1];
    ans[0] = root.val;
    dfs(root, ans);
    return ans[0];
  }
}